src/tokens/reader_stream.rs

```diff
@@ -0,0 +1,560 @@
+    pending_token: PendingToken,
+    position: ReaderPosition,
+    // XXX this allows us to explicitly capture a lifetime for the reader we operate on
+    phantom: PhantomData<&'a R>,
+}
+
+/// These generate the methods that produce lazy content for our stream against the underlying
+/// reader.  There is a lot of boilerplate and variants so this avoids copy/paste errors.
+macro_rules! scalar_content_method {
+    ($name:ident, $scalar_variant:ident, $reader:ident, $read_expr:expr) => {
+        #[inline]
+        fn $name(&mut self) -> Content<'a> {
+            let reader_cell = Rc::clone(&self.reader_cell);
+            ScalarThunk(
+                ScalarType::$scalar_variant,
+                Thunk::defer(move || {
+                    let mut underlying = reader_cell.borrow_mut();
+                    let $reader = underlying.try_mut_ref()?;
+                    Ok(ScalarValue::$scalar_variant($read_expr))
+                }),
+            )
+            .into()
+        }
+    };
+}
+
+impl<'a, R> ReaderTokenStream<'a, R>
+where
+    R: IonReader<Item = StreamItem, Symbol = Symbol> + 'a,
+{
+    #[inline]
+    fn annotations_thunk(&self) -> AnnotationsThunk<'a> {
+        let reader_cell = Rc::clone(&self.reader_cell);
+        Thunk::defer(move || reader_cell.borrow().try_ref()?.annotations().collect())
+    }
+
+    #[inline]
+    fn field_name_thunk(&self) -> FieldNameThunk<'a> {
+        match self.parent_type() {
+            Some(IonType::Struct) => {
+                let reader_cell = Rc::clone(&self.reader_cell);
+                // XXX note that we expect a field name, so we do want that to surface as error
+                //     and not None
+                Thunk::defer(move || Ok(Some(reader_cell.borrow().try_ref()?.field_name()?)))
+            }
+            _ => Thunk::wrap(None),
+        }
+    }
+
+    scalar_content_method!(bool_token, Bool, reader, reader.read_bool()?);
+    scalar_content_method!(int_token, Int, reader, reader.read_int()?);
+    scalar_content_method!(float_token, Float, reader, reader.read_f64()?);
+    scalar_content_method!(decimal_token, Decimal, reader, reader.read_decimal()?);
+    scalar_content_method!(timestamp_token, Timestamp, reader, reader.read_timestamp()?);
+    scalar_content_method!(string_token, String, reader, reader.read_string()?);
+    scalar_content_method!(symbol_token, Symbol, reader, reader.read_symbol()?);
+    scalar_content_method!(blob_token, Blob, reader, reader.read_blob()?.into());
+    scalar_content_method!(clob_token, Clob, reader, reader.read_clob()?.into());
+
+    #[inline]
+    fn invalidate_token(&mut self) {
+        // only invalidate if we have something deferred
+        if self.pending_token.is_deferred() {
+            let underlying = {
+                let mut original = self.reader_cell.borrow_mut();
+                original.invalidate()
+            };
+            self.reader_cell = Rc::new(RefCell::new(underlying));
+        }
+    }
```

Comment:
🗺️ This lets us "garbage collect" outstanding tokens, such tokens will just error when trying to evaluate any deferred value within them.